zstd.cc

```diff
@@ -75,34 +99,47 @@ zstd_processor::zstd_processor(const opt_getter& opts)
     // We assume that the uncompressed input length is always <= chunk_len.
     auto cparams = ZSTD_getCParams(_compression_level, chunk_len, 0);
     auto cctx_size = ZSTD_estimateCCtxSize_usingCParams(cparams);
-    // According to the ZSTD documentation, pointer to the context buffer must be 8-bytes aligned.
-    _cctx_raw = allocate_aligned_buffer<char>(cctx_size, 8);
-    _cctx = ZSTD_initStaticCCtx(_cctx_raw.get(), cctx_size);
-    if (!_cctx) {
-        throw std::runtime_error("Unable to initialize ZSTD compression context");
+
+    if (cctx_size > _shared_cctx_size) [[unlikely]] {
+        // According to the ZSTD documentation, pointer to the context buffer must be 8-bytes aligned.
+        auto new_buf = allocate_aligned_buffer<char>(cctx_size, 8);
+        if (!ZSTD_initStaticCCtx(new_buf.get(), cctx_size)) {
+            // This should never happen.
+            throw std::runtime_error("Unable to initialize ZSTD compression context");
+        }
+        if (cctx_size <= max_shared_cctx_size) {
+            _shared_cctx_size = cctx_size;
+            _shared_cctx = std::move(new_buf);
+        } else {
+            _private_cctx = std::move(new_buf);
+        }
     }
 
-    auto dctx_size = ZSTD_estimateDCtxSize();
-    _dctx_raw = allocate_aligned_buffer<char>(dctx_size, 8);
-    _dctx = ZSTD_initStaticDCtx(_dctx_raw.get(), dctx_size);
-    if (!_cctx) {
-        throw std::runtime_error("Unable to initialize ZSTD decompression context");
+    if (!_shared_dctx) {
+        auto dctx_size = ZSTD_estimateDCtxSize();
+        _shared_dctx = allocate_aligned_buffer<char>(dctx_size, 8);
+        if (!ZSTD_initStaticCCtx(_shared_dctx.get(), dctx_size)) {
+            // This should never happen.
+            throw std::runtime_error("Unable to initialize ZSTD decompression context");
+        }
     }
 }
 
 size_t zstd_processor::uncompress(const char* input, size_t input_len, char* output, size_t output_len) const {
-    auto ret = ZSTD_decompressDCtx(_dctx, output, output_len, input, input_len);
+    auto dctx = reinterpret_cast<ZSTD_DCtx*>(_shared_dctx.get());
```

Comment:
I'm curious why you dropped the "_dctx_raw" vs "_dctx" distinction in the new version of the code, forcing you to do reinterpret_casts below. What's the benefit?

Reply:
No particular reason. It's just that the buffers can be reallocated now, so the typed "observer" pointers would now have to be refreshed / checked for invalidation before use. I thought that reducing the number of variables and always using the shared pointers would be simpler.